### PR TITLE
feat(gitlab): Implement GitLab addReviewers

### LINKED ISSUE #4462: addReviewers not implemented for Gitlab
#### This is a:

- [ ] Bug report (non-security related)
- [x] Feature request
- [ ] I'm not sure which of those it is

#### I'm using:

- [ ] The Renovate GitHub App
- [ ] Self-hosted GitHub
- [x] Self-hosted GitLab
- [ ] Self-hosted Azure DevOps

#### Please describe the issue:
There is a field in the renovate config file for adding `reviewers`. Currently, that feature is unimplemented for Gitlab (see [index.ts:520](https://github.com/renovatebot/renovate/blob/master/lib/platform/gitlab/index.ts#L520)). Gitlab, however, has an API for [adding approvers to an individual merge request](https://docs.gitlab.com/ee/api/merge_request_approvals.html#create-merge-request-level-rule). 

Since the MR approvers are generally equivalent of reviewers in Gitlab, I would like to propose implementing the feature using this `merge_request_approvals` API. You can also enable copying global rules by adding the `approval_project_rule_id`, so I would propose enabling that as well so that users can extend their configured approver lists.

### PR DESCRIPTION
**Pull request by fgblomqvist**
## Changes:

* Implements `addReviewers` for GitLab
* Minor refactoring/method extraction to be a bit more DRY

## Context:

 Closes #4462.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

**Will add tests shortly.**

### CODE REVIEW

File: lib/platform/gitlab/index.ts [hunk 1]
```diff
@@ -839,28 +832,26 @@ export async function ensureIssueClosing(title: string): Promise<void> {
   }
 }
 
+async function getUserID(username: string): Promise<number> {
+  return (
+    await gitlabApi.getJson<{ id: number }[]>(`users?username=${username}`)
+  ).body[0].id;
+}
```
Comment (viceice) [inline lib/platform/gitlab/index.ts:839]: Move to `./http.ts`

File: lib/platform/gitlab/merge-request.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import { logger } from '../../logger';
+import { GitlabHttp } from '../../util/http/gitlab';
+import { GitLabMergeRequest, UpdateMergeRequest } from './types';
+
+export const gitlabApi = new GitlabHttp();
```
Comment (viceice) [inline lib/platform/gitlab/merge-request.ts:5]: move to `./http.ts` and reuse from all files, so single instance.

**Review by viceice — CHANGES_REQUESTED** [review]
Needs jest tests

**Comment by fgblomqvist** [comment]
Ran this on a real test repo, worked well.

**Comment by fgblomqvist** [comment]
Is there something blocking this from getting merged? Just lmk if anything is missing/needs to be done.

File: lib/platform/gitlab/merge-request.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import { logger } from '../../logger';
+import { gitlabApi } from './http';
+import { GitLabMergeRequest, UpdateMergeRequest } from './types';
```
Comment (viceice) [inline lib/platform/gitlab/merge-request.ts:3]:
```suggestion
import type { GitLabMergeRequest, UpdateMergeRequest } from './types';
```

File: lib/platform/gitlab/types.ts [hunk 4]
```diff
@@ -8,6 +8,35 @@ export interface GitlabComment {
   id: number;
 }
 
+export type GitLabUser = {
```
Comment (viceice) [inline lib/platform/gitlab/types.ts:11]: Use type or interface for all, do not mix 🤗

File: lib/platform/gitlab/merge-request.ts (hunk 3 above)
Comment (viceice) [inline lib/platform/gitlab/merge-request.ts:3]:
```suggestion
import type { GitLabMergeRequest, UpdateMergeRequest } from './types';
```

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/platform/gitlab/types.ts (hunk 4 above)
Comment (fgblomqvist) [reply lib/platform/gitlab/types.ts:11]: My thought was that I would use `type` for all types and `interface` for all function parameters, but tbh I've always thought that TS doesn't really make a good enough distinction between types and interfaces so in the past I've always defaulted to just using interfaces everywhere :man_shrugging: .
I guess based on the TS docs: `If you would like a heuristic, use interface until you need to use features from type.`, I'll change these to interfaces.

**Review by viceice — APPROVED** [review]

File: lib/platform/gitlab/index.ts [hunk 5]
```diff
@@ -876,10 +862,51 @@ export async function addAssignees(
...
+    logger.error({ error: err }, 'Failed to get IDs of the new reviewers');
+    return;
+  }
+
+  try {
+    await updateMR(config.repository, iid, {
+      reviewer_ids: [...existingReviewerIDs, ...newReviewerIDs],
+    });
+  } catch (err) {
+    logger.error({ error: err }, 'Failed to add reviewers');
```
Comment (viceice) [inline lib/platform/gitlab/index.ts:908]:
```suggestion
    logger.warn({ err }, 'Failed to add reviewers');
```
Do not fail renovate, just warn

File: lib/platform/gitlab/index.ts [hunk 6]
```diff
@@ -876,10 +862,51 @@ export async function addAssignees(
...
+
+  // Figure out which reviewers (of the ones we want to add) are not already on the MR as a reviewer
+  const newReviewers = reviewers.filter((r) => !existingReviewers.includes(r));
+
+  // Gather the IDs for all the reviewers we want to add
+  let newReviewerIDs: number[];
+  try {
+    newReviewerIDs = await Promise.all(newReviewers.map((r) => getUserID(r)));
+  } catch (err) {
+    logger.error({ error: err }, 'Failed to get IDs of the new reviewers');
```
Comment (viceice) [inline lib/platform/gitlab/index.ts:899]:
```suggestion
    logger.warn({ error: err }, 'Failed to get IDs of the new reviewers');
```

No fail

File: lib/platform/gitlab/index.ts [hunk 7]
```diff
@@ -876,10 +862,51 @@ export async function addAssignees(
...
-  logger.warn('Unimplemented in GitLab: approvals');
-  return Promise.resolve();
+export async function addReviewers(
+  iid: number,
+  reviewers: string[]
+): Promise<void> {
+  logger.debug(`Adding reviewers '${reviewers.join(', ')}' to #${iid}`);
+
+  if (lt(defaults.version, '13.9.0')) {
+    logger.warn(
```
Comment (viceice) [inline lib/platform/gitlab/index.ts:872]:
```suggestion
    logger.debug(
```

No warn, like gitea

File: lib/platform/gitlab/index.ts [hunk 8]
```diff
@@ -876,10 +862,51 @@ export async function addAssignees(
...
+      { currentVersion: defaults.version },
```
Comment (viceice) [inline lib/platform/gitlab/index.ts:873]:
```suggestion
      { version: defaults.version },
```

File: lib/platform/gitlab/index.ts [hunk 9]
```diff
@@ -876,10 +862,51 @@ export async function addAssignees(
...
+      'Adding reviewers is only available in GitLab 13.9 and onwards'
+    );
+    return;
+  }
+
+  let mr: GitLabMergeRequest;
+  try {
+    mr = await getMR(config.repository, iid);
+  } catch (err) {
+    logger.error({ error: err }, 'Failed to get existing reviewers');
```
Comment (viceice) [inline lib/platform/gitlab/index.ts:883]:
```suggestion
    logger.warn({ err }, 'Failed to get existing reviewers');
```

Only warn

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/platform/gitlab/index.ts (hunk 7 above)
Comment (fgblomqvist) [reply lib/platform/gitlab/index.ts:872]: I know gitea does it that way, but doesn't that mean that if someone doesn't have debug logging on, they won't see that reviewers got skipped? And might wonder what went wrong? E.g. if I specify reviewers in my config, I would expect to get at least a warning if it wasn't possible to add them. Or is there something special about the warn handling in renovate? Would info be better?

Comment (viceice) [reply lib/platform/gitlab/index.ts:872]: ok, leave it at warn, we will track it in #9420

File: lib/platform/gitlab/index.spec.ts [hunk 10]
```diff
@@ -44,6 +49,24 @@ describe('platform/gitlab', () => {
     httpMock.reset();
   });
 
+  async function initFakePlatform(version: string) {
+    httpMock.scope(gitlabApiHost).get('/api/v4/user').reply(200, {
+      email: 'a@b.com',
+      name: 'Renovate Bot',
+    });
+    httpMock
+      .scope(gitlabApiHost)
```
Comment (viceice) [inline lib/platform/gitlab/index.spec.ts:58]:
```suggestion
    })
```

reuse same scopes as in other tests

**Review by viceice — COMMENTED** [review]
just a small issue

File: lib/platform/gitlab/index.spec.ts (hunk 10 above)
Comment (fgblomqvist) [reply lib/platform/gitlab/index.spec.ts:58]: Good point, thanks!

File: lib/platform/gitlab/index.ts [hunk 11]
```diff
@@ -876,10 +862,51 @@ export async function addAssignees(
...
+  const existingReviewers = mr.reviewers.map((r) => r.username);
+  const existingReviewerIDs = mr.reviewers.map((r) => r.id);
+
+  // Figure out which reviewers (of the ones we want to add) are not already on the MR as a reviewer
+  const newReviewers = reviewers.filter((r) => !existingReviewers.includes(r));
+
+  // Gather the IDs for all the reviewers we want to add
+  let newReviewerIDs: number[];
+  try {
+    newReviewerIDs = await Promise.all(newReviewers.map((r) => getUserID(r)));
```
Comment (viceice) [inline lib/platform/gitlab/index.ts:897]: We need to rate limit here:
```suggestion
    newReviewerIDs = await pAll(newReviewers.map((r) => () =>getUserID(r)), { concurrency: 5 });
```
Maybe needs another variable
use `import pAll from 'p-all';`

**Review by viceice — CHANGES_REQUESTED** [review]
Sorry, found another issue

File: lib/platform/gitlab/index.ts (hunk 11 above)
Comment (fgblomqvist) [reply lib/platform/gitlab/index.ts]: Note for the future: probably wanna do the rate-limiting inside of the Gitlab query functions (e.g. gitlab.putJson) :smile:

**Review by viceice — APPROVED** [review]

File: lib/platform/gitlab/index.spec.ts [hunk 12]
```diff
@@ -830,14 +854,115 @@ describe('platform/gitlab', () => {
...
+          )
+          .reply(200, { reviewers: existingReviewers })
+          .get('/api/v4/users?username=someuser')
+          .reply(200, [{ id: 1 }])
+          .get('/api/v4/users?username=someotheruser')
+          .reply(200, [{ id: 2 }]);
+
+        await gitlab.addReviewers(42, ['someuser', 'foo', 'someotheruser']);
+        expect(scope.isDone()).toBeTrue();
+      });
```
Comment (viceice) [inline lib/platform/gitlab/index.spec.ts:962]: @fgblomqvist What was your intension about this test

image

Comment (fgblomqvist) [reply lib/platform/gitlab/index.spec.ts:962]: To make sure it didn't make unnecessary API calls? Or what do you mean?

**Pull request merged**